xarray/testing/assertions.py

```diff
@@ -50,21 +52,72 @@ def _data_allclose_or_equiv(arr1, arr2, rtol=1e-05, atol=1e-08, decode_bytes=Tru
 
 
 @ensure_warnings
-def assert_equal(a, b):
+def assert_isomorphic(a: DataTree, b: DataTree, from_root: bool = False):
+    """
+    Two DataTrees are considered isomorphic if every node has the same number of children.
+
+    Nothing about the data in each node is checked.
+
+    Isomorphism is a necessary condition for two trees to be used in a nodewise binary operation,
+    such as tree1 + tree2.
+
+    By default this function does not check any part of the tree above the given node.
+    Therefore this function can be used as default to check that two subtrees are isomorphic.
+
+    Parameters
+    ----------
+    a : DataTree
+        The first object to compare.
+    b : DataTree
+        The second object to compare.
+    from_root : bool, optional, default is False
+        Whether or not to first traverse to the root of the trees before checking for isomorphism.
+        If a & b have no parents then this has no effect.
+
+    See Also
+    --------
+    DataTree.isomorphic
+    assert_equal
+    assert_identical
+    """
+    __tracebackhide__ = True
+    assert isinstance(a, type(b))
+
+    if isinstance(a, DataTree):
+        if from_root:
+            a = a.root
+            b = b.root
+
+        assert a.isomorphic(b, from_root=from_root), diff_datatree_repr(
+            a, b, "isomorphic"
+        )
+    else:
+        raise TypeError(f"{type(a)} not of type DataTree")
+
+
+@ensure_warnings
+def assert_equal(a, b, **kwargs):
     """Like :py:func:`numpy.testing.assert_array_equal`, but for xarray
     objects.
 
     Raises an AssertionError if two objects are not equal. This will match
     data values, dimensions and coordinates, but not names or attributes
     (except for Dataset objects for which the variable names must match).
-    Arrays with NaN in the same location are considered equal.
+    Arrays with NaN in the same location are considered equal. DataTrees are
+    equal if they have isomorphic node structures, with matching node names,
+    and if they have matching variables and coordinates, all of which are
+    equal.
 
     Parameters
     ----------
-    a : xarray.Dataset, xarray.DataArray, xarray.Variable or xarray.Coordinates
-        The first object to compare.
-    b : xarray.Dataset, xarray.DataArray, xarray.Variable or xarray.Coordinates
-        The second object to compare.
+    a : xarray.Dataset, xarray.DataArray, xarray.Variable, xarray.Coordinates
+        or xarray.core.datatree.DataTree. The first object to compare.
+    b : xarray.Dataset, xarray.DataArray, xarray.Variable, xarray.Coordinates
+        or xarray.core.datatree.DataTree. The second object to compare.
+    **kwargs : dict
+        Additional keyword arguments passed on to DataTree.equals. Specifically,
+        from_root, which defaults True, and will check the whole tree above
+        the given node.
```

Comment:
I don't understand why `assert_equal` has `**kwargs` at all - they don't seem to be used?

`assert_allclose` accepts specific named tolerance kwargs, shouldn't we just add `from_root` explicitly to all of these functions?

Reply:
I may have misunderstood the conversation in our tag-up today, but it sounded like `**kwargs` was a potentially preferred option to explicitly stating `from_root`, where `from_root` is only used by `assert_equal` or `assert_identical` when dealing with a `DataTree`. But happy to explicitly name `from_root` as an optional argument if that seems better.

Reply:
Forgetting about `from_root` for a moment, I just don't understand why xarray's existing `assert_equal` method already accepts `**kwargs` then does nothing with them.

Reply:
I'm not sure I follow here - I added `**kwargs` to both `assert_equal` and `assert_identical` in this PR. Then inside the condition for `isinstance(a, DataTree)` I try to grab `from_root` from those `kwargs`.

Reply:
Oops, I'm being stupid, ignore the comment about it already being there (it wasn't).

Why are you adding `**kwargs` though? Did we decide to do that in the meeting? I thought all we decided was that `assert_*` would need a `from_root` arg, even if that didn't do anything for non-`DataTree` inputs.

Reply:
Okay. I can make that change, sounds good.

Reply:
[63e2836](https://github.com/pydata/xarray/pull/8967/commits/63e28362f22881865d2f0c077d02506675d3630a)

Reply:
I've been suggesting that... the idea was that having it for non-`DataTree` objects might be confusing, but I guess `**kwargs` makes tab-completion less useful (but other than that there should not be any difference).

Reply:
Ah my bad for somehow missing that suggestion initially.

Still, I feel like hiding a kwarg away is more confusing than just showing it and saying it is only used if two `DataTree` objects are passed... For example [`apply_ufunc`](https://docs.xarray.dev/en/stable/generated/xarray.apply_ufunc.html#xarray.apply_ufunc) has arguments that are only used if the input is of a certain type.

Another thing we could do here is use `typing.overloads` to ensure that passing `from_root` when `a` and `b` are not both `DataTree`s would fail static type checking.

Reply:
Yeah - I was just thinking about using `typing.overloads` - I'll have a go at that.

Reply:
[970b10](https://github.com/pydata/xarray/pull/8967/commits/970b10d17cf682644031b0aa67043e7aad253ead)